<a href="https://colab.research.google.com/github/CharlesPrado23/Notebooks/blob/main/Semana9/Collaborative_aula(K_Means).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<center>
    <img src="https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/Logos/organization_logo/organization_logo.png" width="300" alt="cognitiveclass.ai logo"  />
</center>

# Collaborative Filtering

Estimated time needed: **25** minutes

## Objectives

After completing this lab you will be able to:

-   Create recommendation system based on collaborative filtering


Recommendation systems are a collection of algorithms used to recommend items to users based on information taken from the user. These systems have become ubiquitous can be commonly seen in online stores, movies databases and job finders. In this notebook, we will explore recommendation systems based on Collaborative Filtering and implement simple version of one using Python and the Pandas library.


<h1>Table of contents</h1>

<div class="alert alert-block alert-info" style="margin-top: 20px">
    <ol>
        <li><a href="#ref1">Acquiring the Data</a></li>
        <li><a href="#ref2">Preprocessing</a></li>
        <li><a href="#ref3">Collaborative Filtering</a></li>
    </ol>
</div>
<br>
<hr>


<a id="ref1"></a>

# Acquiring the Data


To acquire and extract the data, simply run the following Bash scripts:  
Dataset acquired from [GroupLens](http://grouplens.org/datasets/movielens?cm_mmc=Email_Newsletter-_-Developer_Ed%2BTech-_-WW_WW-_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-ML0101EN-SkillsNetwork-20718538&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ&cm_mmc=Email_Newsletter-_-Developer_Ed%2BTech-_-WW_WW-_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-ML0101EN-SkillsNetwork-20718538&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ&cm_mmc=Email_Newsletter-_-Developer_Ed%2BTech-_-WW_WW-_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-ML0101EN-SkillsNetwork-20718538&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ&cm_mmc=Email_Newsletter-_-Developer_Ed%2BTech-_-WW_WW-_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-ML0101EN-SkillsNetwork-20718538&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ). Lets download the dataset. To download the data, we will use **`!wget`** to download it from IBM Object Storage.  
**Did you know?** When it comes to Machine Learning, you will likely be working with large datasets. As a business, where can you host your data? IBM is offering a unique opportunity for businesses, with 10 Tb of IBM Cloud Object Storage: [Sign up now for free](http://cocl.us/ML0101EN-IBM-Offer-CC)


In [1]:
!wget -O moviedataset.zip https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-ML0101EN-SkillsNetwork/labs/Module%205/data/moviedataset.zip
print('unziping ...')
!unzip -o -j moviedataset.zip 

--2021-04-11 00:40:12--  https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-ML0101EN-SkillsNetwork/labs/Module%205/data/moviedataset.zip
Resolving cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud (cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud)... 169.63.118.104
Connecting to cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud (cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud)|169.63.118.104|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 160301210 (153M) [application/zip]
Saving to: ‘moviedataset.zip’

moviedataset.zip    100%[===================>] 152.88M  29.0MB/s    in 5.3s    

2021-04-11 00:40:18 (29.0 MB/s) - ‘moviedataset.zip’ saved [160301210/160301210]

unziping ...
Archive:  moviedataset.zip
  inflating: links.csv               
  inflating: movies.csv              
  inflating: ratings.csv             
  inflating: README.txt              
  inflating: tags.csv        

Now you're ready to start working with the data!


<hr>

<a id="ref2"></a>

# Preprocessing


First, let's get all of the imports out of the way:


In [2]:
#Dataframe manipulation library
import pandas as pd
#Math functions, we'll only need the sqrt function so let's import only that
from math import sqrt
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

Now let's read each file into their Dataframes:


In [3]:
#Storing the movie information into a pandas dataframe
movies_df = pd.read_csv('movies.csv')
#Storing the user information into a pandas dataframe
ratings_df = pd.read_csv('ratings.csv')

Let's also take a peek at how each of them are organized:


In [4]:
#Head is a function that gets the first N rows of a dataframe. N's default is 5.
movies_df.head()

movieId  ...                                       genres
0        1  ...  Adventure|Animation|Children|Comedy|Fantasy
1        2  ...                   Adventure|Children|Fantasy
2        3  ...                               Comedy|Romance
3        4  ...                         Comedy|Drama|Romance
4        5  ...                                       Comedy

[5 rows x 3 columns]

So each movie has a unique ID, a title with its release year along with it (Which may contain unicode characters) and several different genres in the same field. Let's remove the year from the title column and place it into its own one by using the handy [extract](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.extract.html#pandas.Series.str.extract?cm_mmc=Email_Newsletter-_-Developer_Ed%2BTech-_-WW_WW-_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-ML0101EN-SkillsNetwork-20718538&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ&cm_mmc=Email_Newsletter-_-Developer_Ed%2BTech-_-WW_WW-_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-ML0101EN-SkillsNetwork-20718538&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ) function that Pandas has.


Let's remove the year from the **title** column by using pandas' replace function and store in a new **year** column.


In [5]:
#Using regular expressions to find a year stored between parentheses
#We specify the parantheses so we don't conflict with movies that have years in their titles
movies_df['year'] = movies_df.title.str.extract('(\(\d\d\d\d\))',expand=False)
#Removing the parentheses
movies_df['year'] = movies_df.year.str.extract('(\d\d\d\d)',expand=False)
#Removing the years from the 'title' column
movies_df['title'] = movies_df.title.str.replace('(\(\d\d\d\d\))', '')
#Applying the strip function to get rid of any ending whitespace characters that may have appeared
movies_df['title'] = movies_df['title'].apply(lambda x: x.strip())

Let's look at the result!


In [6]:
movies_df.head()

movieId  ...  year
0        1  ...  1995
1        2  ...  1995
2        3  ...  1995
3        4  ...  1995
4        5  ...  1995

[5 rows x 4 columns]

With that, let's also drop the genres column since we won't need it for this particular recommendation system.


In [7]:
#Dropping the genres column
movies_df = movies_df.drop('genres', 1)

Here's the final movies dataframe:


In [8]:
movies_df.head()

movieId                        title  year
0        1                    Toy Story  1995
1        2                      Jumanji  1995
2        3             Grumpier Old Men  1995
3        4            Waiting to Exhale  1995
4        5  Father of the Bride Part II  1995

<br>


Next, let's look at the ratings dataframe.


In [9]:
ratings_df.head()

userId  movieId  rating   timestamp
0       1      169     2.5  1204927694
1       1     2471     3.0  1204927438
2       1    48516     5.0  1204927435
3       2     2571     3.5  1436165433
4       2   109487     4.0  1436165496

Every row in the ratings dataframe has a user id associated with at least one movie, a rating and a timestamp showing when they reviewed it. We won't be needing the timestamp column, so let's drop it to save on memory.


In [10]:
#Drop removes a specified row or column from a dataframe
ratings_df = ratings_df.drop('timestamp', 1)

Here's how the final ratings Dataframe looks like:


In [11]:
ratings_df.head()

userId  movieId  rating
0       1      169     2.5
1       1     2471     3.0
2       1    48516     5.0
3       2     2571     3.5
4       2   109487     4.0

<hr>

<a id="ref3"></a>

# Collaborative Filtering


Now, time to start our work on recommendation systems. 

The first technique we're going to take a look at is called **Collaborative Filtering**, which is also known as **User-User Filtering**. As hinted by its alternate name, this technique uses other users to recommend items to the input user. It attempts to find users that have similar preferences and opinions as the input and then recommends items that they have liked to the input. There are several methods of finding similar users (Even some making use of Machine Learning), and the one we will be using here is going to be based on the **Pearson Correlation Function**.

<img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-ML0101EN-SkillsNetwork/labs/Module%205/images/User_Item.png" width=800px>

The process for creating a User Based recommendation system is as follows:

-   Select a user with the movies the user has watched
-   Based on his rating to movies, find the top X neighbours 
-   Get the watched movie record of the user for each neighbour.
-   Calculate a similarity score using some formula
-   Recommend the items with the highest score

Let's begin by creating an input user to recommend movies to:

Notice: To add more movies, simply increase the amount of elements in the userInput. Feel free to add more in! Just be sure to write it in with capital letters and if a movie starts with a "The", like "The Matrix" then write it in like this: 'Matrix, The' .


In [78]:
userInput = [
            {'title':'Breakfast Club, The', 'rating':1.5},
            {'title':'Toy Story', 'rating':3.5},
            {'title':'Jumanji', 'rating':4.5},
            {'title':"Pulp Fiction", 'rating':2.5},
            {'title':'Akira', 'rating':5}
         ] 
inputMovies = pd.DataFrame(userInput)
inputMovies

title  rating
0  Breakfast Club, The     1.5
1            Toy Story     3.5
2              Jumanji     4.5
3         Pulp Fiction     2.5
4                Akira     5.0

#### Add movieId to input user

With the input complete, let's extract the input movies's ID's from the movies dataframe and add them into it.

We can achieve this by first filtering out the rows that contain the input movies' title and then merging this subset with the input dataframe. We also drop unnecessary columns for the input to save memory space.


In [79]:
#Filtering out the movies by title
inputId = movies_df[movies_df['title'].isin(inputMovies['title'].tolist())]
#Then merging it so we can get the movieId. It's implicitly merging it by title.
inputMovies = pd.merge(inputId, inputMovies)
#Dropping information we won't use from the input dataframe
inputMovies = inputMovies.drop('year', 1)
#Final input dataframe
#If a movie you added in above isn't here, then it might not be in the original 
#dataframe or it might spelled differently, please check capitalisation.
inputMovies

movieId                title  rating
0        1            Toy Story     3.5
1        2              Jumanji     4.5
2      296         Pulp Fiction     2.5
3     1274                Akira     5.0
4     1968  Breakfast Club, The     1.5

#### The users who has seen the same movies

Now with the movie ID's in our input, we can now get the subset of users that have watched and reviewed the movies in our input.


In [113]:
#Filtering out users that have watched movies that the input has watched and storing it
userSubset = ratings_df[ratings_df['movieId'].isin(inputMovies['movieId'].tolist())]
userSubset


userId  movieId  rating
19             4      296     4.0
441           12     1968     3.0
479           13        2     2.0
531           13     1274     5.0
681           14      296     2.0
...          ...      ...     ...
22883679  247738      296     4.0
22884132  247751        1     4.0
22884142  247751      296     4.0
22884164  247751     1274     5.0
22884355  247753        1     3.0

[196623 rows x 3 columns]

We now group up the rows by user ID.


In [81]:
#Groupby creates several sub dataframes where they all have the same value in the column specified as the parameter
userSubsetGroup = userSubset.groupby(['userId'])

lets look at one of the users, e.g. the one with userID=1130


In [82]:
userSubsetGroup.get_group(1130)
len(userSubsetGroup)

116140

Let's also sort these groups so the users that share the most movies in common with the input have higher priority. This provides a richer recommendation since we won't go through every single user.


In [83]:
#Sorting it so users with movie most in common with the input will have priority
userSubsetGroup = sorted(userSubsetGroup,  key=lambda x: len(x[1]), reverse=True)

Now lets look at the first user


In [84]:
userSubsetGroup = userSubsetGroup[0:1200]

In [85]:
userSubsetGroup[0:5]


[(75,       userId  movieId  rating
  7507      75        1     5.0
  7508      75        2     3.5
  7540      75      296     5.0
  7633      75     1274     4.5
  7673      75     1968     5.0), (106,       userId  movieId  rating
  9083     106        1     2.5
  9084     106        2     3.0
  9115     106      296     3.5
  9198     106     1274     3.0
  9238     106     1968     3.5), (686,        userId  movieId  rating
  61336     686        1     4.0
  61337     686        2     3.0
  61377     686      296     4.0
  61478     686     1274     4.0
  61569     686     1968     5.0), (815,        userId  movieId  rating
  73747     815        1     4.5
  73748     815        2     3.0
  73922     815      296     5.0
  74362     815     1274     3.0
  74678     815     1968     4.5), (1040,        userId  movieId  rating
  96689    1040        1     3.0
  96690    1040        2     1.5
  96733    1040      296     3.5
  96859    1040     1274     3.0
  96922    1040     1968  

# K-Means - Modelo de Clusterização

In [86]:
userId = userSubsetGroup[100][0]
grades = userSubsetGroup[100][1].rating.tolist()
columns = userSubsetGroup[0][1].movieId.tolist()
#print(userId)
#print(grades)
print(columns)

[1, 2, 296, 1274, 1968]


In [87]:
## user_id Movie_id(0) ..... movie_id(4)
##.  75       5                  10
##.  80.      1                   3
w_dict= {} 

for k in range(len(userSubsetGroup)):
  Id = userSubsetGroup[k][0]
  grades = userSubsetGroup[k][1].rating.tolist()
  w_dict[Id] = grades 

#columns = userSubsetGroup[0][1].movieId.tolist()

df = pd.DataFrame.from_dict(w_dict,orient='index')
df.columns = columns 
df['userId'] = df.index
df.index = range(len(df))



In [88]:
df.head()

1    2  296  1274  1968  userId
0  5.0  3.5  5.0   4.5   5.0      75
1  2.5  3.0  3.5   3.0   3.5     106
2  4.0  3.0  4.0   4.0   5.0     686
3  4.5  3.0  5.0   3.0   4.5     815
4  3.0  1.5  3.5   3.0   4.0    1040

In [89]:
X = df[columns]


In [90]:
from sklearn.cluster import KMeans 
kmeans = KMeans(n_clusters=2)
kmeans.fit(X)

KMeans(algorithm='auto', copy_x=True, init='k-means++', max_iter=300,
       n_clusters=2, n_init=10, n_jobs=None, precompute_distances='auto',
       random_state=None, tol=0.0001, verbose=0)

In [91]:
from sklearn.metrics import silhouette_score

for i in range(2,10):
  kmeans_ = KMeans(n_clusters=i).fit(X)
  labels_ = kmeans_.labels_
  coef = silhouette_score(X,labels_)
  print("N_cluster: {}, score: {}".format(i,coef))

N_cluster: 2, score: 0.22532787189828601
N_cluster: 3, score: 0.21472558208348147
N_cluster: 4, score: 0.16514017120648627
N_cluster: 5, score: 0.15032004831105827
N_cluster: 6, score: 0.1573208815559621
N_cluster: 7, score: 0.1663911744587207
N_cluster: 8, score: 0.1483896602822771
N_cluster: 9, score: 0.15628606178114182


In [92]:
y_labels = kmeans.predict(X)
n = np.unique(y_labels,return_counts=True)
n

(array([0, 1], dtype=int32), array([488, 712]))

In [93]:
### A que grupo pertence o usuário input_user ### 
x_input = np.array(inputMovies.rating.tolist())
x_input = x_input.reshape(1,5)
x_output = kmeans.predict(x_input)
x_output

array([0], dtype=int32)

In [94]:
y_labels

array([1, 0, 1, ..., 0, 1, 1], dtype=int32)

In [95]:
X_dist = kmeans.transform(X)
X_dist

array([[3.09782036, 1.31869663],
       [1.14840196, 2.47221643],
       [2.09144804, 1.16788733],
       ...,
       [4.74581084, 6.0049301 ],
       [1.49629844, 1.13495291],
       [1.97322633, 1.25847693]])

In [96]:
dist_ = np.min(X_dist,axis=1)
dist_
idx_similarity = 1/dist_
idx_similarity


array([0.75832453, 0.87077525, 0.85624698, ..., 0.21071215, 0.88109382,
       0.79461131])

In [97]:
df_user = pd.DataFrame()
df_user['similarityIndex'] = idx_similarity
df_user['userId'] = df['userId']
df_user['clusterLabels'] = y_labels
df_user.head()

similarityIndex  userId  clusterLabels
0         0.758325      75              1
1         0.870775     106              0
2         0.856247     686              1
3         0.719690     815              1
4         0.731462    1040              0

In [98]:
df_user = df_user[df_user['clusterLabels']==x_output[0]]
df_user = df_user.drop('clusterLabels',axis=1)

In [99]:
df_user.head()

similarityIndex  userId
1          0.870775     106
4          0.731462    1040
5          0.272588    1130
10         1.365845    2065
12         0.693872    2432

#### The top x similar users to input user

Now let's get the top 100 users that are most similar to the input.


In [100]:
topUsers=df_user.sort_values(by='similarityIndex', ascending=False)[0:100]
topUsers.head()

similarityIndex  userId
246          2.610905   42734
1101         2.610905  193846
534          2.549328   90904
799          2.549328  142163
270          1.696340   46987

Now, let's start recommending movies to the input user.

#### Rating of selected users to all movies

We're going to do this by taking the weighted average of the ratings of the movies using the Pearson Correlation as the weight. But to do this, we first need to get the movies watched by the users in our **pearsonDF** from the ratings dataframe and then store their correlation in a new column called _similarityIndex". This is achieved below by merging of these two tables.


In [101]:
topUsersRating=topUsers.merge(ratings_df, left_on='userId', right_on='userId', how='inner')
topUsersRating.head()

similarityIndex  userId  movieId  rating
0         2.610905   42734        1     3.5
1         2.610905   42734        2     2.5
2         2.610905   42734        6     2.0
3         2.610905   42734       10     2.5
4         2.610905   42734       19     2.0

Now all we need to do is simply multiply the movie rating by its weight (The similarity index), then sum up the new ratings and divide it by the sum of the weights.

We can easily do this by simply multiplying two columns, then grouping up the dataframe by movieId and then dividing two columns:

It shows the idea of all similar users to candidate movies for the input user:


In [102]:
#Multiplies the similarity by the user's ratings
topUsersRating['weightedRating'] = topUsersRating['similarityIndex']*topUsersRating['rating']
topUsersRating.head()

similarityIndex  userId  movieId  rating  weightedRating
0         2.610905   42734        1     3.5        9.138166
1         2.610905   42734        2     2.5        6.527261
2         2.610905   42734        6     2.0        5.221809
3         2.610905   42734       10     2.5        6.527261
4         2.610905   42734       19     2.0        5.221809

In [103]:
#Applies a sum to the topUsers after grouping it up by userId
tempTopUsersRating = topUsersRating.groupby('movieId').sum()[['similarityIndex','weightedRating']]
tempTopUsersRating.columns = ['sum_similarityIndex','sum_weightedRating']
tempTopUsersRating.head()

sum_similarityIndex  sum_weightedRating
movieId                                         
1                 111.955146          383.819356
2                 111.955146          288.308505
3                  35.139478           96.154069
4                   3.682987           10.171045
5                  14.446704           29.090198

In [104]:
#Creates an empty dataframe
recommendation_df = pd.DataFrame()
#Now we take the weighted average
recommendation_df['weighted average recommendation score'] = tempTopUsersRating['sum_weightedRating']/tempTopUsersRating['sum_similarityIndex']
recommendation_df['movieId'] = tempTopUsersRating.index
recommendation_df.head()

weighted average recommendation score  movieId
movieId                                                
1                                     3.428332        1
2                                     2.575214        2
3                                     2.736354        3
4                                     2.761630        4
5                                     2.013622        5

Now let's sort it and see the top 20 movies that the algorithm recommended!


In [105]:
recommendation_df = recommendation_df.sort_values(by='weighted average recommendation score', ascending=False)
recommendation_df.head()

weighted average recommendation score  movieId
movieId                                                
54326                                      5.0    54326
1493                                       5.0     1493
3223                                       5.0     3223
5042                                       5.0     5042
6408                                       5.0     6408

In [106]:
movies_df.loc[movies_df['movieId'].isin(recommendation_df.head(20)['movieId'].tolist())]

movieId                                              title  year
596        602                             Great Day in Harlem, A  1994
1451      1493                        Love and Other Catastrophes  1996
3137      3223                               Zed & Two Noughts, A  1985
3945      4038                                      Kill Me Again  1989
4947      5042                                     Forbidden Zone  1980
5261      5357                                          Iron Will  1994
6301      6408                       Animals are Beautiful People  1974
7056      7167                                     Japanese Story  2003
7351      7488                               Berkeley in the '60s  1990
7421      7699  Who the Hell Is Juliette? (¿Quién diablos es J...  1997
7653      8120                                        29th Street  1991
9181     27033                         Kingdom II, The (Riget II)  1997
9370     27513                               Dog Days (Hundstage)  2001
12027    54326                                         Sierra, La  2005
12032    54354                                         China Blue  2005
12136    55207                                           Cashback  2004
13014    61913                                       Africa addio  1966
14057    70465       Jerusalema (Gangster's Paradise: Jerusalema)  2008
15652    79541                 Little Nicholas (Le petit Nicolas)  2009
17558    88203                              Car Bonus (Autobonus)  2001

### Advantages and Disadvantages of Collaborative Filtering

##### Advantages

-   Takes other user's ratings into consideration
-   Doesn't need to study or extract information from the recommended item
-   Adapts to the user's interests which might change over time

##### Disadvantages

-   Approximation function can be slow
-   There might be a low of amount of users to approximate
-   Privacy issues when trying to learn the user's preferences


<h2>Want to learn more?</h2>

IBM SPSS Modeler is a comprehensive analytics platform that has many machine learning algorithms. It has been designed to bring predictive intelligence to decisions made by individuals, by groups, by systems – by your enterprise as a whole. A free trial is available through this course, available here: <a href="https://www.ibm.com/analytics/spss-statistics-software">SPSS Modeler</a>

Also, you can use Watson Studio to run these notebooks faster with bigger datasets. Watson Studio is IBM's leading cloud solution for data scientists, built by data scientists. With Jupyter notebooks, RStudio, Apache Spark and popular libraries pre-packaged in the cloud, Watson Studio enables data scientists to collaborate on their projects without having to install anything. Join the fast-growing community of Watson Studio users today with a free account at <a href="https://www.ibm.com/cloud/watson-studio">Watson Studio</a>


### Thank you for completing this lab!

## Author

Saeed Aghabozorgi

### Other Contributors

<a href="https://www.linkedin.com/in/joseph-s-50398b136/" target="_blank">Joseph Santarcangelo</a>

## Change Log

| Date (YYYY-MM-DD) | Version | Changed By | Change Description                 |
| ----------------- | ------- | ---------- | ---------------------------------- |
| 2020-11-03        | 2.1     | Lakshmi    | Updated URL of csv                 |
| 2020-08-27        | 2.0     | Lavanya    | Moved lab to course repo in GitLab |
|                   |         |            |                                    |
|                   |         |            |                                    |

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>
